src/main/java/bio/terra/cli/service/utils/GoogleCloudStorage.java

```diff
@@ -74,4 +79,33 @@ public void deleteBucket(String bucketUri) {
       throw new RuntimeException("Bucket deletion failed.");
     }
   }
+
+  /**
+   * Check whether we have permission to list objects in the bucket.
+   *
+   * <p>List access is included in the Storage Object Viewer and Storage Legacy Bucket Reader. It is
+   * NOT included in the Storage Legacy Object Reader.
+   *
+   * @param bucketUri uri of the bucket (gs://...)
+   * @return true if access is allowed
+   */
+  public boolean checkObjectsListAccess(String bucketUri) {
+    try {
+      String bucketName = bucketUri.replaceFirst("^gs://", "");
+
+      // try listing objects in the bucket
+      storageClient.list(
+          bucketName,
+          Storage.BlobListOption.userProject(googleProjectId),
+          Storage.BlobListOption.fields());
+      return true;
+    } catch (StorageException storageEx) {
```

Comment:
[Not necessarily for this PR]

I went through a round of testing and got tripped up here. I was using a workspace that I'd created maybe 1 or 2 weeks ago. And I kept getting "Invalid or inaccessible bucket path" exceptions when trying to create a data-ref, even after double-checking that my access was set up properly.

After turning on debug-level logging in logback.xml, I was able to see the underlying error:

`com.google.cloud.storage.StorageException: gjuggler@gmail.com does not have serviceusage.services.use access to the Google Cloud project.`

Aha – this is because we're always passing in the "context" project ID to the GCS list call, which is required for requester-pays buckets. However, a user must have the serviceusage.services.use permission on this project in order for a call to succeed ([ref](https://cloud.google.com/storage/docs/requester-pays#requirements)).

It seems clear that https://github.com/DataBiosphere/terra-workspace-manager/pull/230 should fix this for newer workspaces. But I thought it's worth noting that this is needed not just for Nextflow, but for any use of the workspace context as a GCS requester-pays billing project.

On a secondary note, it would have been nice to somehow get a more specific error message here. Two possible improvements to make:

(1) Add a user-accessible --debug or --verboase flag that shows some of the logger logs. This would make it easier for me to debug without having to recompile. Maybe worth filing a low-priority follow up ticket?
(2) Potentially increase the specificity of the user-facing message on the access-check. In this case, it might make more sense to inspect the StorageException a bit more and re-throw a more user-friendly exception message, rather than a simple true or false, since there may be multiple causes of a failed access-check.

Reply:
Good point about the requester-pays buckets. Yes, that's why I'm passing the project into the GCS class, but I hadn't tested it on an older workspace.

I created a new [ticket](https://broadworkbench.atlassian.net/browse/PF-542) to control logging without requiring a re-compile. Prefer to hold off on better exception interpretation for now, until we get some feedback from the solutions team about what `check-access` should mean.